EQL: sequence - add support for missing events

luigidellaquila opened the pull request:
eg. the following query 

```
sequence with maxspan=10m
   [ event where name = 'a' ]
  ![ event where name = 'b' ]
   [ event where name = 'c' ]
```

will return results if there are two events "a" and "c" in a time window of 10 minutes, but there is not event "b" between the two.

The syntax `![ <condition> ]` identifies a missing event criterion, ie. a criterion that does not have to match any events for the sequence to be valid.

Missing events are supported as first criteria, as last criteria and/or as criteria in-between normal (positive) events.

Queries with missing events for now have to define a "maxspan"

luigidellaquila committed:
Add grammar for EQL sequences with missing events

luigidellaquila committed:
Implement in-between missing events management

luigidellaquila committed:
Add support for leading and trailing missing events

luigidellaquila committed:
Remove unused code

luigidellaquila committed:
Fix support for 'by key' and add further tests

luigidellaquila committed:
Add missing event placeholders to the result

and fix query composition for missing events

luigidellaquila committed:
Remove leftover from tests

luigidellaquila commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into eql/missing_events_simple

luigidellaquila committed:
Remove unused code

luigidellaquila committed:
Merge remote-tracking branch 'luigidellaquila/eql/missing_events_simple' into eql/missing_events_simple

elasticsearchmachine commented:
Hi @luigidellaquila, I've created a changelog YAML for you.

elasticsearchmachine commented:
Pinging @elastic/es-ql (Team:QL)

luigidellaquila committed:
Update docs/changelog/92348.yaml

astefan commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/assembler/ExecutionManager.java:
```diff
@@ -214,6 +224,23 @@ public Executable assemble(List<List<Attribute>> listOfKeys, List<PhysicalPlan>
         return new SampleIterator(new PITAwareQueryClient(session), criteria, cfg.fetchSize(), limit, session.circuitBreaker());
     }
 
+    private int indexOfFirstPositive(boolean[] missing) {
+        for (int i = 0; i < missing.length; i++) {
+            if (missing[i] == false) {
+                return i;
+            }
+        }
+        throw new IllegalStateException("Invalid sequence query: at least one event has to be positive");
+    }
```
I think this entire check needs to happen at parsing time, likely in the LogicalPlanBuilder.

luigidellaquila replied:
I agree with you about throwing the exception at parsing time, I'm adding it to LogicalPlanBuilder.
I don't like very much the idea of calculating the index of the first missing event at parsing time and passing it to the constructor, mostly because it would allow to pass conflicting information (ie. a `boolean[] missing` and the wrong `indexOfFirstPositive`), so I'd prefer to keep the calculation here, maybe doing it only once at the beginning.

astefan commented on x-pack/plugin/eql/src/main/antlr/EqlBase.g4:
```diff
@@ -212,6 +212,7 @@ LP: '(';
 RP: ')';
 PIPE: '|';
 OPTIONAL: '?';
+MISSING_EVENT_OPEN: '-[';
```
I know we've discussed this over and over again, and I've re-read now everything around this subject. While I find it more logical and syntax appropriate to use a suffix for the missing events feature, in combination with the `with` construct (think of `with runs=2, missing=true`) there are too many drawback to this approach and one compelling advantage to a prefix alternative, to push `with missing` forward. As such, I think we should use a prefix (like you suggested in this PR). One detail (that you already implemented) is that we don't have to have whitespaces between the "missing" character and the opening square bracket.

Also, I tend to prefer `!`. Minus sign has too many other implications imo. So the syntax should be `sequence by k1 with maxspan=1s [ test5 where tag == \"normal\" ] by k2 with runs=2 ![ test5 where tag == \"missing\" ] by k2 [ test5 where tag == \"normal\" ] by k2`

luigidellaquila replied:
The minus was just a temporary decision that allowed us to proceed with the implementation without waiting for an agreement on the final syntax.
I'm absolutely fine with `![`, I'm doing the change now.

astefan commented on x-pack/plugin/eql/qa/common/src/main/resources/test_missing_events.toml:
```diff
@@ -0,0 +1,340 @@
+   -[ test1 where tag == "nonexisting"]
+    [ test1 where tag == "normal" ]
+'''
+expected_event_ids = [1, -1, 2, 2, -1, 4]
```
```suggestion
expected_event_ids = [1, -1, 2,
                      2, -1, 4]
```

astefan commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/SequenceMatcher.java:
```diff
@@ -218,18 +239,88 @@ private void match(int stage, SequenceKey key, Ordinal ordinal, HitReference hit
+        return stage == firstPositiveStage();
+    }
+
+    int firstPositiveStage() {
```
Instead of this method called from various places every time and unless I'm missing something regarding an update to `missingEvents`, firstPositiveStage can be a simple `int`. The same goes for `lastPositiveStage`.

astefan commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/SequenceMatcher.java:
```diff
@@ -72,11 +72,13 @@ public void clear() {
     // Set of completed sequences - separate to avoid polluting the other stages
     // It is a set since matches are ordered at insertion time based on the ordinal of the first entry
     private final Set<Sequence> completed;
+    private final Set<Sequence> toCheckForMissing;
```
I think this one can be considered for clearing up in `clear()` method as well, no?

astefan commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/SequenceMatcher.java:
```diff
@@ -293,6 +384,14 @@ void trim(Ordinal ordinal) {
+        this.completed.add(sequence);
+    }
+
+    public Set<Sequence> toCheckForMissing() {
```
```suggestion
    Set<Sequence> toCheckForMissing() {
```

astefan commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/SequenceMatcher.java:
```diff
@@ -293,6 +384,14 @@ void trim(Ordinal ordinal) {
+        return toCheckForMissing;
+    }
+
     public Stats stats() {
         return stats;
     }
```
This method is not used I think. Can be removed.

astefan commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/TumblingWindow.java:
```diff
@@ -57,7 +76,11 @@
+        new SearchModule(Settings.EMPTY, List.of()).getNamedWriteables()
+    );
     private static final int CACHE_MAX_SIZE = 64;
+    private static final int MAX_SEQUENCES_TO_CHECK_FOR_MISSING = 1000;
```
Please, add a comment regarding the purpose of this limit.

astefan commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/SequenceMatcher.java:
```diff
@@ -218,18 +239,88 @@ private void match(int stage, SequenceKey key, Ordinal ordinal, HitReference hit
     }
 
+    public void tryComplete(Sequence sequence) {
+        for (int i = 0; i < missingEvents.length; i++) {
+            if (missingEvents[i]) {
+                toCheckForMissing.add(sequence);
+                return;
+            }
+        }
```
I read this as "if there is any missing event filter in this sequence, then check the current sequence". This could easily be a boolean (where if there is at least one `true` value in `missingEvents`), no need to recheck `missingEvents` in a `for` loop every time.

astefan commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/assembler/BoxedQueryRequest.java:
```diff
@@ -57,6 +58,7 @@ public BoxedQueryRequest(QueryRequest original, String timestamp, List<String> k
         searchSource = original.searchSource();
         // setup range queries and preserve their reference to simplify the update
         timestampRange = rangeQuery(timestamp).timeZone("UTC").format("epoch_millis");
```
```suggestion
        timestampRange = timestampRangeQuery(timestamp);
```

astefan commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/assembler/BoxedQueryRequest.java:
```diff
@@ -74,6 +76,10 @@ public void nextAfter(Ordinal ordinal) {
+    public String timestampField() {
+        return timestampField;
+    }
+
```
```suggestion
    public RangeQueryBuilder timestampRangeQuery() {
        return timestampRangeQuery(timestampField);
    }

    private static RangeQueryBuilder timestampRangeQuery(String timestamp) {
        return rangeQuery(timestamp).timeZone("UTC").format("epoch_millis");
    }
```

astefan commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/TumblingWindow.java:
```diff
@@ -161,15 +192,166 @@ private void tumbleWindow(int currentStage, ActionListener<Payload> listener) {
+                SequenceCriterion criterion = criteria.get(i);
+                if (criterion.missing()) {
+                    BoxedQueryRequest r = criterion.queryRequest();
+                    RangeQueryBuilder range = new RangeQueryBuilder(r.timestampField()).timeZone("UTC").format("epoch_millis");
```
Externalize this range query builder to `BoxedQueryRequest`.
```suggestion
                    RangeQueryBuilder range = RangeQueryBuilder range = r.timestampRangeQuery();
```

astefan commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/TumblingWindow.java:
```diff
@@ -161,15 +192,166 @@ private void tumbleWindow(int currentStage, ActionListener<Payload> listener) {
+                            continue;
+                        }
+                        Timestamp hitTimestamp = criterion.timestamp(hits[0]);
+                        lastLeading = lastLeading == null || lastLeading.delta(hitTimestamp) < 0 ? hitTimestamp : lastLeading;
```
Why the choice of `delta` method here? You don't actually need the delta (which is calculated differently) but if one is earlier than the other. Which means a `compareTo` is enough: `lastLeading.instant().compareTo(hitTimestamp.instant())`. There are slight differences between two approaches, but if I'm not missing something the compareTo above is still using the nanos precision.

astefan replied:
This one is not reflected in the changes.

astefan commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/TumblingWindow.java:
```diff
@@ -161,15 +192,166 @@ private void tumbleWindow(int currentStage, ActionListener<Payload> listener) {
+        checkMissingEvents(next, listener);
+    }
+
+    private boolean biggerThanMaxSpan(Timestamp from, Timestamp to) {
```
`SequenceMatcher.exceedsMasSpan` is doing almost the same thing. My suggestion is to externalize this behavior and make it generic so that both places can use it. Also, in SequenceMatcher the `delta` is used which seems the better choice.

astefan commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/TumblingWindow.java:
```diff
@@ -161,15 +192,166 @@ private void tumbleWindow(int currentStage, ActionListener<Payload> listener) {
+                if ((firstTrailing == null && lastLeading == null)
+                    || (lastLeading == null && matcher.isMissingEvent(0))
+                    || (firstTrailing == null && matcher.isMissingEvent(lastStage))
+                    || (matcher.isMissingEvent(0) && matcher.isMissingEvent(lastStage) && biggerThanMaxSpan(lastLeading, firstTrailing))
```
👍

astefan commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/TumblingWindow.java:
```diff
@@ -161,15 +192,166 @@ private void tumbleWindow(int currentStage, ActionListener<Payload> listener) {
+                }
+            }
+        }
+        checkMissingEvents(next, listener);
```
Why are we calling this here? This potentially calls again `doCheckMissingEvents` which will have the same sequences to check... but `sequencesToCheck` Set should probably be empty if there were less sequences in the set than MAX_SEQUENCES_TO_CHECK_FOR_MISSING, but if there were more than MAX_SEQUENCES_TO_CHECK_FOR_MISSING then `sequencesToCheck` still has sequences to check for missing and this is when these are checked. This means that MAX_SEQUENCES_TO_CHECK_FOR_MISSING is a way to "paginate" the search for missing events through the set of gathered sequences?

luigidellaquila replied:
Yes, that's the idea. Since sequences can be pretty big (up to 255 terms if I'm not mistaken), each term can be a missing event to check, and we will check multiple sequences with a single request, the multi-query could grow significantly.

astefan replied:
Imo, the name `MAX_SEQUENCES_TO_CHECK_FOR_MISSING` is not descriptive enough, probably that's why I also suggested https://github.com/elastic/elasticsearch/pull/92348#discussion_r1190856136. The current name suggests this is an absolute maximum setting (like `LogicalPlanBuilder.MAX_SAMPLE_QUERIES` for example), but in fact it's a page size kind of check: "this number of sequences can be checked at a time". Maybe you can find a more descriptive name for it? `MISSING_EVENTS_SEQUENCES_CHECK_BATCH_SIZE`? (I think `BATCH_SIZE` is a good choice, not too fond of the rest of the name though but it's not bad). Wdyt?

luigidellaquila replied:
I like the "BATCH_SIZE" 👍

astefan commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/TumblingWindow.java:
```diff
@@ -161,15 +192,166 @@ private void tumbleWindow(int currentStage, ActionListener<Payload> listener) {
+    /*
+     * Not a great way of getting a copy of a SearchSourceBuilder
+     */
+    private static SearchSourceBuilder copySource(SearchSourceBuilder source) {
```
For this method, please create an `ExecutionUtils` class under `org.elasticsearch.xpack.eql.execution` package and add it there (if you have other ideas/options please let me know). Replace the call in SampleQueryRequest, as well.

astefan commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/TumblingWindow.java:
```diff
@@ -229,6 +411,11 @@ private void baseCriterion(int baseStage, SearchResponse r, ActionListener<Paylo
         } else {
             info = null;
+            // this covers the case where there is only one positive criterion and all the others are missing events
+            if (baseStage == matcher.firstPositiveStage() && base.descending()) {
```
I don't see why those two checks cover the scenario described in the comment: "only one positive criterion and all the others are missing events".

luigidellaquila replied:
Yes, there is something wrong (in the checks, not in the scenario). Let me review them.

astefan commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/TumblingWindow.java:
```diff
@@ -499,7 +695,7 @@ private void boxQuery(WindowInfo window, SequenceCriterion criterion) {
     private void setupWindowFromTail(Ordinal from) {
         // TAIL can only be at stage 0
         // the ASC window starts at stage 1
```
This comment should reflect the below change

astefan commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/TumblingWindow.java:
```diff
@@ -511,7 +707,7 @@ private void setupWindowFromTail(Ordinal from) {
             }
             // reset all sub queries
-            for (int i = 2; i < maxStages; i++) {
+            for (int i = nextPositiveStage(matcher.firstPositiveStage()); i < maxStages; i++) {
```
Is this accurate? The original algorithm used the first criterion as the base query sorted descending, while the rest of the criteria sorted ascending. The original algorithm reset criteria starting with 2, in your case you start from the same positive stage not from the next next positive stage.

luigidellaquila committed:
Merge branch 'main' into eql/missing_events_simple

luigidellaquila committed:
Merge remote-tracking branch 'luigidellaquila/eql/missing_events_simple' into eql/missing_events_simple

luigidellaquila committed:
Replace -[ syntax with ![

eg.

sequence with maxspan=10m
   [ event where name = 'a' ]
  ![ event where name = 'b' ]
   [ event where name = 'c' ]

luigidellaquila committed:
Implement review suggestions

luigidellaquila committed:
Merge branch 'main' into eql/missing_events_simple

luigidellaquila committed:
Implement review suggestions

luigidellaquila commented:
@elasticmachine run elasticsearch-ci/bwc

luigidellaquila committed:
Merge branch 'main' into eql/missing_events_simple

luigidellaquila committed:
Revert changes on Timestamp

luigidellaquila commented:
@elasticmachine run elasticsearch-ci/part-2

luigidellaquila committed:
Better management of range queries to check missing events

astefan commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/parser/LogicalPlanBuilder.java:
```diff
@@ -330,11 +334,31 @@ public Sequence visitSequence(SequenceContext ctx) {
         }
 
+        if (maxSpan.duration() < 0 && queries.stream().anyMatch(x -> x.missingEvent())) {
+            throw new ParsingException(
+                source,
+                "No maxspan defined. Maxspan is mandatory for sequences with missing events",
+                queries.size()
+            );
```
The style of parsing error message is a bit different. I've attempted a rewording for this one. Also `queries.size()` is not needed.

```suggestion
            throw new ParsingException(
                source,
                "[maxspan] is required for sequences with missing events queries; found none"
            );
```

astefan commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/parser/LogicalPlanBuilder.java:
```diff
@@ -330,11 +334,31 @@ public Sequence visitSequence(SequenceContext ctx) {
+        }
+
+        if (queries.stream().allMatch(KeyedFilter::missingEvent)) {
+            throw new IllegalStateException("Invalid sequence query: at least one event has to be positive");
```
```suggestion
            throw new IllegalStateException("A sequence requires at least one positive event query; found none");
```

astefan commented on x-pack/plugin/eql/qa/common/src/main/resources/test_missing_events.toml:
```diff
@@ -0,0 +1,387 @@
+   ![ test1 where tag == "nonexisting"]
+    [ test1 where tag == "normal" ]
+'''
+expected_event_ids = [1, -1, 2,
```
Wdyt about changing the `-` sign of the results to `!` as well? To me, it feels a bit more intuitive when looking over the tests...

luigidellaquila replied:
Do you mean in the `expected_event_ids`? It's a `long[]`, that's why I used a negative value (I cannot even use null)

astefan replied:
Yeah, nevermind. I forgot about the pains and limitations on working with toml.

astefan commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/SequenceMatcher.java:
```diff
@@ -218,18 +273,68 @@ private void match(int stage, SequenceKey key, Ordinal ordinal, HitReference hit
+        return headLimit;
+    }
+
+    public long maxSpanInNanos() {
```
This one is not used.

astefan commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/plan/logical/KeyedFilter.java:
```diff
@@ -29,22 +29,31 @@ public class KeyedFilter extends UnaryPlan {
     private final Attribute tiebreaker;
-
-    public KeyedFilter(Source source, LogicalPlan child, List<? extends NamedExpression> keys, Attribute timestamp, Attribute tiebreaker) {
+    private final boolean missingEvent;
```
I would rename this to be a bit more descriptive and a bit more accurate: `isMissingEventFilter`.

astefan commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/SequenceMatcher.java:
```diff
@@ -72,11 +73,16 @@ public void clear() {
     private final boolean descending;
 
     private final Limit limit;
+    private final boolean[] missingEvents;
```
Since the algorithm is complex enough, a bit more descriptive naming can help.
I would suggest a rename for this to `missingEventStages` (singular `event` and plural `stage`). In essence we are not talking about the actual events but about "stages" or "queries" or "criteria" (I am open to suggestions on which of these three words describe best this concept) that refer to matching missing events.

astefan commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/SequenceMatcher.java:
```diff
@@ -72,11 +73,16 @@ public void clear() {
+    private final boolean[] missingEvents;
+    protected final int firstPositiveStage;
+    protected final int lastPositiveStage;
+    private final boolean missingEventsExist;
```
`missingEventQueriesExist`
`missingEventStagesExist`
`missingEventFiltersExist`
`missingEventCriteriaExist` (less favorable on this one because the "criteria" is used in other classes, this one is more about "stages")

luigidellaquila replied:
I like `missingEventStagesExist`

astefan commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/sequence/SequenceMatcher.java:
```diff
@@ -72,11 +73,16 @@ public void clear() {
 
     private final Limit limit;
+    private final boolean[] missingEvents;
+    protected final int firstPositiveStage;
```
The "positive" and "negative" meaning needs to reflect in the documentation as well, since we have error messages that refer to them.

luigidellaquila replied:
I double-checked the error messages and everything seems OK.
We'll have to update the docs when this PR is final.

luigidellaquila committed:
Merge branch 'main' into eql/missing_events_simple

luigidellaquila commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/assembler/ExecutionManager.java:
```diff
@@ -129,7 +137,8 @@ public Executable assemble(
                     tbExtractor,
                     itbExtractor,
-                    i == 0 && descending
+                    i == firstPositive && descending,
```
I checked the implementation and I think the redundancy is in the TumblingWindow. The original check in particular, practically just checked that whether first criterion was descending. 
The second criterion was always ascending, so  `criteria.get(0).descending() != criteria.get(1).descending()` 
was practically equivalent to `criteria.get(0).descending()`.

With missing events, that check can be replaced with `criteria.get(matcher.firstPositiveStage).descending()`.

The code in ExecutionManager is correct, the first positive criterion is the query that has to be evaluated in descending order in case of TAIL queries; we cannot change it, or we'll lose the asc/desc information

luigidellaquila committed:
Implement review suggestions

astefan reviewed (approved):
LGTM
Let's wait a bit more before merging, for the hands-on testing to finish.

luigidellaquila commented:
@elasticmachine run elasticsearch-ci/part-3

elasticmachine committed:
Merge branch 'main' into eql/missing_events_simple

luigidellaquila commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into eql/missing_events_simple
